midenc-compile/src/stages/cross_ctx/lower_imports.rs

```diff
@@ -0,0 +1,281 @@
+    if dfg.get_import(&import_func_id).is_none() {
+        dfg.import_function(
+            import_func_id.module,
+            import_func_id.function,
+            import_lowered_sig.clone(),
+        )
+        .map_err(|_e| {
+            let message = format!(
+                "lowering function with name {} in module {} with signature \
+                 {import_lowered_sig:?} is already imported (function call) with a different \
+                 signature",
+                import_func_id.function, import_func_id.module
+            );
+            diagnostics.diagnostic(Severity::Error).with_message(message).into_report()
+        })?;
+    }
+    let call = builder.ins().call(import_func_id, &params, span);
+    // dbg!(&sig);
+    let result = builder.first_result(call);
+    builder.ins().ret(Some(result), span);
+    let function_id = builder.build()?;
+    module_builder.build()?;
+    let component_import = MidenAbiImport::new(FunctionType {
+        abi: Abi::Canonical,
+        params: import_lowered_sig.params.into_iter().map(|p| p.ty).collect(),
+        results: import_lowered_sig.results.into_iter().map(|r| r.ty).collect(),
+    });
+    // dbg!(&component_import);
+    // dbg!(&function_id);
+
+    Ok((component_import, function_id))
+}
+
+/// Rewrites the calls to the function `from` to the function `to` in the given function.
+/// Returns `true` if any call was rewritten.
+fn rewrite_calls(function: &mut Function, from: FunctionIdent, to: FunctionIdent) -> bool {
+    let mut dirty = false;
+    let mut worklist = VecDeque::<Block>::default();
+    for block in function.dfg.blocks.keys() {
+        worklist.push_back(block);
+    }
+
+    for b in worklist {
+        let block = &mut function.dfg.blocks[b];
+        // Take the list of instructions away from the block to simplify traversing the block
+        let mut insts = block.insts.take();
+        // Take each instruction out of the list, top to bottom, modify it, then
+        // add it back to the instruction list of the block directly. This ensures
+        // we traverse the list and rewrite instructions in a single pass without
+        // any additional overhead
+        while let Some(inst) = insts.pop_front() {
+            let mut inst = unsafe { UnsafeRef::into_box(inst) };
+            let instruction: &mut Instruction = inst.as_mut();
+            match instruction {
+                Instruction::Call(Call { ref mut callee, .. }) => {
+                    // Rewrite the call instruction
+                    // rewrite_call(call, &mut function.dfg.value_lists, rewrites);
+                    if callee == &from {
+                        *callee = to;
+                        dirty = true;
+                    }
+                }
+                _op => (),
+            }
+
+            block.insts.push_back(UnsafeRef::from_box(inst));
+        }
+    }
+    dirty
+}
+
+/// Replaces calls to the imported functions with calls to the lowering functions.
```

Comment:
Not sure about the current IR, but in HIR2 I expect we can avoid needing to touch the calls in the core module at all, since whether a function is a declaration (import) or a definition solely depends on whether it has a body or not. When we synthesize the body, none of the callers are affected.

I thought that was largely the same in the current IR, but my memory is a bit fuzzy on that detail.

Reply:
I intend to avoid replacing calls in the new HIR2 frontend. Here I'm rewriting them since the core Wasm import(function) in the note project for the `canon lower` has the same name (e.g. `miden:cross-ctx-account/foo@1.0.0#process-felt`) as the synthetic function generated for the `canon lift` in the account package, so I decided to rename the `canon lower` one (hardcoded prefix). In the HIR2 frontend, I plan to handle this in the core module import parsing, renaming them on the fly.